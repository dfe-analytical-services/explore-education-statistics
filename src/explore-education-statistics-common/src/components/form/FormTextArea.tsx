import FormGroup from '@common/components/form/FormGroup';
import FormLabel, { FormLabelProps } from '@common/components/form/FormLabel';
import classNames from 'classnames';
import React, {
  ChangeEventHandler,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
} from 'react';
import ErrorMessage from '../ErrorMessage';

export interface FormTextAreaProps extends FormLabelProps {
  className?: string;
  error?: ReactNode | string;
  hint?: string;
  id: string;
  maxLength?: number;
  name: string;
  rows?: number;
  value?: string;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  onClick?: MouseEventHandler<HTMLTextAreaElement>;
  onKeyPress?: KeyboardEventHandler<HTMLTextAreaElement>;
}

const FormTextArea = ({
  className,
  error,
  hint,
  id,
  hideLabel,
  label,
  maxLength,
  rows = 5,
  value,
  ...props
}: FormTextAreaProps) => {
  const textArea = (
    <>
      <FormLabel id={id} label={label} hideLabel={hideLabel} />

      {hint && (
        <span id={`${id}-hint`} className="govuk-hint">
          {hint}
        </span>
      )}

      {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}

      <textarea
        {...props}
        aria-describedby={
          classNames({
            [`${id}-error`]: !!error,
            [`${id}-hint`]: !!hint,
            [`${id}-info`]: !!maxLength,
          }) || undefined
        }
        className={classNames('govuk-textarea', className, {
          'govuk-js-character-count govuk-textarea--error':
            maxLength && (value?.length ?? 0) > maxLength,
        })}
        id={id}
        rows={rows}
        value={value}
      />
    </>
  );

  if (!!maxLength && maxLength > 0) {
    const remaining = maxLength - (value?.length ?? 0);

    return (
      <div className="govuk-character-count">
        <FormGroup>{textArea}</FormGroup>

        <span
          aria-live="polite"
          className={classNames('govuk-character-count__message', {
            'govuk-hint': remaining >= 0,
            'govuk-error-message': remaining < 0,
          })}
          id={`${id}-info`}
        >
          {remaining >= 0
            ? `You have ${remaining} character${
                remaining !== 1 ? 's' : ''
              } remaining`
            : `You have ${Math.abs(remaining)} character${
                remaining !== -1 ? 's' : ''
              } too many`}
        </span>
      </div>
    );
  }

  return textArea;
};

export default FormTextArea;
