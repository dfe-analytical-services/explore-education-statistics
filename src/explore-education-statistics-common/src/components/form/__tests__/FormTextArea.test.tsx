import { render } from '@testing-library/react';
import noop from 'lodash/noop';
import React from 'react';
import FormTextArea from '../FormTextArea';

describe('FormTextArea', () => {
  test('renders correctly with required props', () => {
    const { container, getByLabelText } = render(
      <FormTextArea id="test-input" label="Test input" name="testInput" />,
    );

    expect(getByLabelText('Test input')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with hint', () => {
    const { container, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
      />,
    );

    const hint = getByText('Fill me in');

    expect(hint.id).toBe('test-input-hint');
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with error', () => {
    const { container, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        error="Field is required"
        onChange={noop}
      />,
    );

    const error = getByText('Field is required');

    expect(error.id).toBe('test-input-error');
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('aria-describedby is equal to the hint id', () => {
    const { getByLabelText, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        onChange={noop}
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-hint',
    );
  });

  test('aria-describedby is equal to the error id', () => {
    const { getByLabelText, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        error="Field is required"
        onChange={noop}
      />,
    );

    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );
    expect(getByLabelText('Test input')).toHaveAttribute(
      'aria-describedby',
      'test-input-error',
    );
  });

  test('aria-describedby contains both hint and error ids', () => {
    const { getByLabelText, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        hint="Fill me in"
        error="Field is required"
      />,
    );

    expect(getByText('Fill me in')).toHaveAttribute('id', 'test-input-hint');
    expect(getByText('Field is required')).toHaveAttribute(
      'id',
      'test-input-error',
    );

    const ariaDescribedBy = getByLabelText('Test input').getAttribute(
      'aria-describedby',
    );

    expect(ariaDescribedBy).toContain('test-input-error');
    expect(ariaDescribedBy).toContain('test-input-hint');
  });

  test('shows a character count message when `maxLength` is above 0', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={10}
      />,
    );

    expect(queryByText('You have 10 characters remaining')).not.toBeNull();
  });

  test('aria-describedby contains the character count message id when `maxLength` is above 0', () => {
    const { getByLabelText, getByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={10}
      />,
    );

    const ariaDescribedBy = getByLabelText('Test input').getAttribute(
      'aria-describedby',
    );

    expect(getByText('You have 10 characters remaining')).toHaveAttribute(
      'id',
      'test-input-info',
    );
    expect(ariaDescribedBy).toContain('test-input-info');
  });

  test('does not show a character count message when `maxLength` is below 0', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={-1}
      />,
    );

    expect(queryByText(/You have .+ characters remaining/)).toBeNull();
  });

  test('does not show a character count message when `maxLength` is 0', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={0}
      />,
    );

    expect(queryByText(/You have .+ characters remaining/)).toBeNull();
  });

  test('shows correct character count message when difference to `maxLength` is 1', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={4}
        value="aaa"
        onChange={noop}
      />,
    );

    expect(queryByText('You have 1 character remaining')).not.toBeNull();
  });

  test('shows correct character count message when difference to `maxLength` is 0', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={4}
        value="aaaa"
        onChange={noop}
      />,
    );

    expect(queryByText('You have 0 characters remaining')).not.toBeNull();
  });

  test('shows correct character count message when difference to `maxLength` is -1', () => {
    const { queryByText } = render(
      <FormTextArea
        id="test-input"
        label="Test input"
        name="testInput"
        maxLength={4}
        value="aaaaa"
        onChange={noop}
      />,
    );

    expect(queryByText('You have 1 character too many')).not.toBeNull();
  });
});
