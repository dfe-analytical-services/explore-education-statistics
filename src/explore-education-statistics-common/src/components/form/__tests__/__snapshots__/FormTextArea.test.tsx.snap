// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextArea renders correctly with error 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span class="govuk-error-message"
      id="test-input-error"
>
  Field is required
</span>
<textarea name="testInput"
          aria-describedby="test-input-error"
          class="govuk-textarea"
          id="test-input"
          rows="5"
>
</textarea>
`;

exports[`FormTextArea renders correctly with hint 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<span id="test-input-hint"
      class="govuk-hint"
>
  Fill me in
</span>
<textarea name="testInput"
          aria-describedby="test-input-hint"
          class="govuk-textarea"
          id="test-input"
          rows="5"
>
</textarea>
`;

exports[`FormTextArea renders correctly with required props 1`] = `
<label class="govuk-label"
       for="test-input"
>
  Test input
</label>
<textarea name="testInput"
          class="govuk-textarea"
          id="test-input"
          rows="5"
>
</textarea>
`;
